### ISSUE TITLE
@extender enhancements

### ISSUE THREAD

**Issue by craigsapp** [rism-digital/verovio#3646]
In this example:

<img width="584" alt="Screenshot 2024-04-14 at 1 30 21 PM" src="https://github.com/rism-digital/verovio/assets/3487289/e6e0bacb-9d94-4979-bb80-c18a1d524744">

<details>
<summary> Click to view MEI data for above example. </summary>

```xml
<?xml version="1.0" encoding="UTF-8"?>
<?xml-model href="https://music-encoding.org/schema/5.0/mei-all.rng" type="application/xml" schematypens="http://relaxng.org/ns/structure/1.0"?>
<?xml-model href="https://music-encoding.org/schema/5.0/mei-all.rng" type="application/xml" schematypens="http://purl.oclc.org/dsdl/schematron"?>
<mei xmlns="http://www.music-encoding.org/ns/mei" meiversion="5.0">
 <meiHead>
  <fileDesc>
   <titleStmt>
    <title />
   </titleStmt>
  </fileDesc>
  <encodingDesc>
   <appInfo>
    <application isodate="2024-04-09T06:42:58" version="4.2.0-dev-84f99b0-dirty">
     <name>Verovio</name>
     <p>Transcoded from Humdrum</p>
    </application>
   </appInfo>
  </encodingDesc>
  <extMeta>
   <frames xmlns="http://www.humdrum.org/ns/humxml" />
  </extMeta>
 </meiHead>
 <music decls="#work1_encoded">
  <body>
   <mdiv xml:id="m1ancjid">
    <score xml:id="smdmgt3">
     <scoreDef xml:id="s1lkx7zi" midi.bpm="400.000000">
      <staffGrp xml:id="s8grm0f">
       <staffDef xml:id="staffdef-L1F1" n="1" lines="5">
        <clef xml:id="cpbtnkz" shape="G" line="2" />
       </staffDef>
      </staffGrp>
     </scoreDef>
     <section xml:id="section-L1F1">
      <measure xml:id="measure-L1">
       <staff xml:id="staff-L1F1" n="1">
        <layer xml:id="layer-L1F1N1" n="1">
         <note xml:id="note-L2F1" dur="1" oct="4" pname="g" accid.ges="n" />
        </layer>
       </staff>
       <dynam xml:id="dynam-L2F2" staff="1" tstamp="1.000000" vgrp="100">f</dynam>
      </measure>
      <measure xml:id="measure-L3">
       <staff xml:id="staff-L3F1N1" n="1">
        <layer xml:id="layer-L3F1N1" n="1">
         <note xml:id="note-L5F1" dur="1" oct="4" pname="a" accid.ges="n" />
        </layer>
       </staff>
       <dynam xml:id="dynam-L5F2" staff="1" tstamp="1.000000" tstamp2="3m+1">
           cresc.
       </dynam>
      </measure>
      <measure xml:id="measure-L6">
       <staff xml:id="staff-L6F1N1" n="1">
        <layer xml:id="layer-L6F1N1" n="1">
         <note xml:id="note-L7F1" dur="1" oct="4" pname="b" accid.ges="n" />
        </layer>
       </staff>
      </measure>
      <measure xml:id="measure-L8">
       <staff xml:id="staff-L8F1N1" n="1">
        <layer xml:id="layer-L8F1N1" n="1">
         <note xml:id="note-L9F1" dur="1" oct="5" pname="c" accid.ges="n" />
        </layer>
       </staff>
      </measure>
      <measure xml:id="measure-L10">
       <staff xml:id="staff-L10F1N1" n="1">
        <layer xml:id="layer-L10F1N1" n="1">
         <note xml:id="note-L11F1" dur="1" oct="5" pname="d" accid.ges="n" />
        </layer>
       </staff>
       <dynam xml:id="dynam-L11F2" staff="1" tstamp="1.000000" vgrp="100">ff</dynam>
      </measure>
     </section>
    </score>
   </mdiv>
  </body>
 </music>
</mei>
```

</details>

The *cresc.* `<dynam>`  with only `@tstamp2` is generating the extension line:

```xml
       <dynam xml:id="dynam-L5F2" staff="1" tstamp="1.000000" tstamp2="3m+1">
          cresc.
       </dynam>
```

Preferably `@extender="true"` should be required for the extension line to be shown:

```xml
       <dynam xml:id="dynam-L5F2" staff="1" tstamp="1.000000" extender="true" tstamp2="3m+1">
          cresc.
       </dynam>
```

This is because the *cresc.* can have a duration without having an extension line after it. (i.e., The *cresc.* has 
a duration of three measures, so for MIDI rendering this information is useful regardless of the visibility of an extension line, and useful for performance rendering the *cresc.*.

But interestingly, `@extender` is being ignored, since the line is still displayed when `@extender="false"`:

```xml
       <dynam xml:id="dynam-L5F2" staff="1" tstamp="1.000000" extender="false" tstamp2="3m+1">
          cresc.
       </dynam>
```

The only way to suppress the extender line is to remove `@tstamp2`.

------------------

In any case, the main feature request is to implement `@endid` for line extensions (specifically in `<dynam>` but also for other elements that can have `@extender` such as `<dir>`).

In other words:

```xml
       <dynam xml:id="dynam-L5F2" staff="1" tstamp="1.000000" extender="true" endid="#dynam-L11F2">
          cresc.
       </dynam>
```

where `dynam-L11F2` is the ID of the `ff` dynamic in measure 5.   Curretly, when `@endid` is used instead of `@tstamp2`, no line extender is given even if `@extender="true"` is specified:

<img width="579" alt="Screenshot 2024-04-14 at 1 43 24 PM" src="https://github.com/rism-digital/verovio/assets/3487289/05588c66-23ae-4d93-8e13-1cebf08ad3f5">

<details>
<summary> Click to view MEI data for above example. </summary>

```xml
<?xml version="1.0" encoding="UTF-8"?>
<?xml-model href="https://music-encoding.org/schema/5.0/mei-all.rng" type="application/xml" schematypens="http://relaxng.org/ns/structure/1.0"?>
<?xml-model href="https://music-encoding.org/schema/5.0/mei-all.rng" type="application/xml" schematypens="http://purl.oclc.org/dsdl/schematron"?>
<mei xmlns="http://www.music-encoding.org/ns/mei" meiversion="5.0">
 <meiHead>
  <fileDesc>
   <titleStmt>
    <title />
   </titleStmt>
  </fileDesc>
  <encodingDesc>
   <appInfo>
    <application isodate="2024-04-09T06:42:58" version="4.2.0-dev-84f99b0-dirty">
     <name>Verovio</name>
     <p>Transcoded from Humdrum</p>
    </application>
   </appInfo>
  </encodingDesc>
  <extMeta>
   <frames xmlns="http://www.humdrum.org/ns/humxml" />
  </extMeta>
 </meiHead>
 <music decls="#work1_encoded">
  <body>
   <mdiv xml:id="m1ancjid">
    <score xml:id="smdmgt3">
     <scoreDef xml:id="s1lkx7zi" midi.bpm="400.000000">
      <staffGrp xml:id="s8grm0f">
       <staffDef xml:id="staffdef-L1F1" n="1" lines="5">
        <clef xml:id="cpbtnkz" shape="G" line="2" />
       </staffDef>
      </staffGrp>
     </scoreDef>
     <section xml:id="section-L1F1">
      <measure xml:id="measure-L1">
       <staff xml:id="staff-L1F1" n="1">
        <layer xml:id="layer-L1F1N1" n="1">
         <note xml:id="note-L2F1" dur="1" oct="4" pname="g" accid.ges="n" />
        </layer>
       </staff>
       <dynam xml:id="dynam-L2F2" staff="1" tstamp="1.000000" vgrp="100">f</dynam>
      </measure>
      <measure xml:id="measure-L3">
       <staff xml:id="staff-L3F1N1" n="1">
        <layer xml:id="layer-L3F1N1" n="1">
         <note xml:id="note-L5F1" dur="1" oct="4" pname="a" accid.ges="n" />
        </layer>
       </staff>
       <dynam xml:id="dynam-L5F2" staff="1" tstamp="1.000000" extender="true" endid="#dynam-L11F2">
    cresc.
       </dynam>
      </measure>
      <measure xml:id="measure-L6">
       <staff xml:id="staff-L6F1N1" n="1">
        <layer xml:id="layer-L6F1N1" n="1">
         <note xml:id="note-L7F1" dur="1" oct="4" pname="b" accid.ges="n" />
        </layer>
       </staff>
      </measure>
      <measure xml:id="measure-L8">
       <staff xml:id="staff-L8F1N1" n="1">
        <layer xml:id="layer-L8F1N1" n="1">
         <note xml:id="note-L9F1" dur="1" oct="5" pname="c" accid.ges="n" />
        </layer>
       </staff>
      </measure>
      <measure xml:id="measure-L10">
       <staff xml:id="staff-L10F1N1" n="1">
        <layer xml:id="layer-L10F1N1" n="1">
         <note xml:id="note-L11F1" dur="1" oct="5" pname="d" accid.ges="n" />
        </layer>
       </staff>
       <dynam xml:id="dynam-L11F2" staff="1" tstamp="1.000000" vgrp="100">ff</dynam>
      </measure>
     </section>
    </score>
   </mdiv>
  </body>
 </music>
</mei>
```

</details>

Here the *cresc.* is:

```xml
       <dynam xml:id="dyna-L5F2" staff="1" tstamp="1.000000" extender="true" endid="#dynam-L11F2">
          cresc.
      </dynam>
```

-------------------

A secondary enhancement is that when `@endid` is used to terminate the extender, the origin element of the extender and the `@endid` element should automatically be placed in the same `@vgrp`.   The exception would be if both elements have `@vgrp` specified, or if the `@staff` or `@place` are difference (for some reason).   This would not apply to notes/rests but only to `<dir>`, `<dynam>`, `<tempo>` (i.e., most or many control elements).

Ideally the previous example would be rendered like this:

<img width="527" alt="Screenshot 2024-04-14 at 1 54 33 PM" src="https://github.com/rism-digital/verovio/assets/3487289/184372a7-f827-4d3e-827a-151f17e3b82a">

<details>
<summary> Click to view MEI data for above example. </summary>

```xml
<?xml version="1.0" encoding="UTF-8"?>
<?xml-model href="https://music-encoding.org/schema/5.0/mei-all.rng" type="application/xml" schematypens="http://relaxng.org/ns/structure/1.0"?>
<?xml-model href="https://music-encoding.org/schema/5.0/mei-all.rng" type="application/xml" schematypens="http://purl.oclc.org/dsdl/schematron"?>
<mei xmlns="http://www.music-encoding.org/ns/mei" meiversion="5.0">
 <meiHead>
  <fileDesc>
   <titleStmt>
    <title />
   </titleStmt>
  </fileDesc>
  <encodingDesc>
   <appInfo>
    <application isodate="2024-04-09T06:42:58" version="4.2.0-dev-84f99b0-dirty">
     <name>Verovio</name>
     <p>Transcoded from Humdrum</p>
    </application>
   </appInfo>
  </encodingDesc>
  <extMeta>
   <frames xmlns="http://www.humdrum.org/ns/humxml" />
  </extMeta>
 </meiHead>
 <music decls="#work1_encoded">
  <body>
   <mdiv xml:id="m1ancjid">
    <score xml:id="smdmgt3">
     <scoreDef xml:id="s1lkx7zi" midi.bpm="400.000000">
      <staffGrp xml:id="s8grm0f">
       <staffDef xml:id="staffdef-L1F1" n="1" lines="5">
        <clef xml:id="cpbtnkz" shape="G" line="2" />
       </staffDef>
      </staffGrp>
     </scoreDef>
     <section xml:id="section-L1F1">
      <measure xml:id="measure-L1">
       <staff xml:id="staff-L1F1" n="1">
        <layer xml:id="layer-L1F1N1" n="1">
         <note xml:id="note-L2F1" dur="1" oct="4" pname="g" accid.ges="n" />
        </layer>
       </staff>
       <dynam xml:id="dynam-L2F2" staff="1" tstamp="1.000000" vgrp="100">f</dynam>
      </measure>
      <measure xml:id="measure-L3">
       <staff xml:id="staff-L3F1N1" n="1">
        <layer xml:id="layer-L3F1N1" n="1">
         <note xml:id="note-L5F1" dur="1" oct="4" pname="a" accid.ges="n" />
        </layer>
       </staff>
       <dynam xml:id="dynam-L5F2" staff="1" vgrp="123" tstamp="1.000000" extender="true" tstamp2="3m+1">
    cresc.
       </dynam>
      </measure>
      <measure xml:id="measure-L6">
       <staff xml:id="staff-L6F1N1" n="1">
        <layer xml:id="layer-L6F1N1" n="1">
         <note xml:id="note-L7F1" dur="1" oct="4" pname="b" accid.ges="n" />
        </layer>
       </staff>
      </measure>
      <measure xml:id="measure-L8">
       <staff xml:id="staff-L8F1N1" n="1">
        <layer xml:id="layer-L8F1N1" n="1">
         <note xml:id="note-L9F1" dur="1" oct="5" pname="c" accid.ges="n" />
        </layer>
       </staff>
      </measure>
      <measure xml:id="measure-L10">
       <staff xml:id="staff-L10F1N1" n="1">
        <layer xml:id="layer-L10F1N1" n="1">
         <note xml:id="note-L11F1" dur="1" oct="3" pname="a" accid.ges="n" />
        </layer>
       </staff>
       <dynam xml:id="dynam-L11F2" vgrp="123" staff="1" tstamp="1.000000">ff</dynam>
      </measure>
     </section>
    </score>
   </mdiv>
  </body>
 </music>
</mei>
```

</details>

where the *cresc.* and `ff` are given matching `@vgrp` explicitly:

```xml
       <dynam xml:id="dynam-L5F2" staff="1" vgrp="123" tstamp="1.000000" extender="true" tstamp2="3m+1">
         cresc.
       </dynam>

       <dynam xml:id="dynam-L11F2" vgrp="123" staff="1" tstamp="1.000000">ff</dynam>
```

The enhancement that would generate the same rendering with an implicit matching `@vgrp`:

```xml
      <dynam xml:id="dynam-L5F2" staff="1" tstamp="1.000000" extender="true" endid="#dynam-L11F2">
         cresc.
      </dynam>

      <dynam xml:id="dynam-L11F2" staff="1" tstamp="1.000000">ff</dynam>
```

Implementing `dynam@endid` would also facilitate MIDI performance renderings regardless of an extender line.

**Comment by rettinghaus** [comment]
Your first issue is solved by fixing a condition in #3648. Your second one isn't really an issue. 
Verovio doesn't track non-layer IDs, so it simply cannot find your given `endid`. Instead of pointing to the next `dynam` you must point to the end note. 

:warning: If you want to say the next `dynam` is actuallly the next `dynam` you should use `@next` on `dynam`.

**Comment by craigsapp** [comment]
Thanks for enhancement no. 1!

For `@next` I suppose that is not used for terminating line extenders (I will check later).  It would be useful for automatic vertical alignment of the two `<dynam>` elements that are attached by a line, since they are never (or very rarely not in horizontal alignment). I don't have a problem with adding them explicitly since I convert from Humdrum, but for efficiency of hand encoding in MEI, this would be preferrable.
